_posts/2016-08-10-dynamic-loading.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Dynamic Code Loading on a MCU
+authors:
+  - brghena
+  - alevy
+---
+
+One key feature of Tock is the ability to load and run multiple applications
+simultaneously. In a modern computer, the OS uses the Memory Management Unit
+(MMU) to provide a virtual address space for each process. The code for each
+process is written assuming that it is the only code in the world and that it
+can place memory at any address it pleases. The MMU handles translating these
+virtual addresses into a physical address in real memory, which is shared
+between all processes. Unfortunately, in the world of embedded systems MMUs are
+not available. Processors like the ARM Cortex-M series omit them because they
+aren't necessary for traditional single-process embedded devices.
+
+A common approach to handle this issue is to assign addresses to applications
+at compile-time. For example, Application A can be placed at address 0x21000
+and Application B can be placed at address 0x22000. This runs into problems
+when Application A grows in size in the future. Moreover, with Tock we want to
+be able to handle dynamically adding, updating, and removing applications at
+run time. Assigning each an address in advance simply isn't possible.
+
+In Tock, we use position independent code (PIC) [^1] to enable loading
+multiple applications. In PIC, all branches and jumps are PC-relative
+rather than absolute, allowing code to be placed at any address in
+Flash. All references to the data section are indirected through the
+Global Offset Table (GOT) [^2]. Rather than access data at an absolute
+address, first the address of the data is loaded from a hard-coded
+offset into the GOT, and then the data is accessed at that address. This
+allows the OS to simply relocate all addresses in the GOT at load time
+based on the actual location of SRAM rather than fixing various
+instructions throughout the code. The address of the GOT itself is
+stored in a PIC base register which is set by the OS before switching to
+application code and is different for each application.
+
+While PIC handles the majority of addressing issues, it does not fix
+everything. Data members which are themselves pointers are assigned a
+value by the compiler rather than indirecting through the GOT. For
+example, take the statement:
+
+```c
+  char* str = "Hello";
```

Comment:
It actually does not result in a warning (surprisingly), but it _is_ incorrect:

```c
char *str = "Hello";

void main() {
  str[0] = 'h';
}
```

segfaults, since it is still placed in rodata. you'd expect it to just be placed in non-`rodata` instead.

However, the problem is not _limited_ to data placed in static sections. Any data that's _reference_ in statically allocated variables is a problem:

```c
int foo = 1234;
struct Bar { int* x} bar = { .x = &foo };
```

Results in the same problem.